presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java

```diff
@@ -0,0 +1,227 @@
+        String queryId = runner.executeWithQueryId(session, query).getQueryId();
+        QueryPlan queryPlan = runner.getQueryPlan(new QueryId(queryId));
+        QueryStats queryStats = runner.getQueryInfo(new QueryId(queryId)).getQueryStats();
+
+        List<PlanNode> allPlanNodes = new PlanNodeSearcher(queryPlan.getPlan().getRoot()).findAll();
+
+        System.out.println(format("Query TPCH [%s] produces [%s] plan nodes.\n", queryNumber, allPlanNodes.size()));
+
+        List<MetricComparison> comparisons = getMetricComparisons(queryPlan, queryStats);
+
+        Map<String, Map<MetricComparison.Result, List<MetricComparison>>> metricSummaries =
+                comparisons.stream().collect(groupingBy(e -> e.metricName, groupingBy(MetricComparison::result)));
+
+        metricSummaries.forEach((metricName, resultSummaries) -> {
+            System.out.println(format("Summary for metric [%s]", metricName));
+            outputSummary(resultSummaries, NO_ESTIMATE);
+            outputSummary(resultSummaries, NO_BASELINE);
+            outputSummary(resultSummaries, DIFFER);
+            outputSummary(resultSummaries, MATCH);
+            System.out.println();
+        });
+
+        System.out.println("Detailed results:\n");
+
+        comparisons.forEach(System.out::println);
+    }
+
+    private List<MetricComparison> getMetricComparisons(QueryPlan queryPlan, QueryStats queryStats)
+    {
+        return METRICS.entrySet().stream().flatMap(metricEntry -> {
+            String metricName = metricEntry.getKey();
+            Function<PlanNodeCost, Estimate> metricFunction = metricEntry.getValue();
+            Map<PlanNode, PlanNodeCost> estimatesByPlanNode = getCostEstimatesByPlanNode(queryPlan);
+            return estimatesByPlanNode.entrySet().stream().map(nodeCostEntry -> {
+                PlanNode node = nodeCostEntry.getKey();
+                PlanNodeCost estimate = nodeCostEntry.getValue();
+                Map<PlanNodeId, PlanNodeCost> executionCostsByPlanId = getExecutionCostsById(queryStats);
+                Optional<PlanNodeCost> execution = Optional.ofNullable(executionCostsByPlanId.get(node.getId()));
+                Optional<Double> estimatedCost = asOptional(metricFunction.apply(estimate));
+                Optional<Double> executionCost = execution.flatMap(e -> asOptional(metricFunction.apply(e)));
+                return new MetricComparison(node, metricName, estimatedCost, executionCost);
+            });
+        }).collect(Collectors.toList());
+    }
+
+    private Map<PlanNode, PlanNodeCost> getCostEstimatesByPlanNode(QueryPlan queryPlan)
+    {
+        return queryPlan.getPlanNodeCosts().entrySet().stream()
+                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private Map<PlanNodeId, PlanNodeCost> getExecutionCostsById(QueryStats queryStats)
+    {
+        //for some reason, there are duplicate PlanNodeId-s in operatorSummaries.
```

Comment:
There might be multiple operators created for single `PlanNodeId`. Example: for join, both `LookupJoinOperator` and `HashBuilderOperator` share the same `planNodeId`. In that case, I don't think you can assume costs will be always be same for same plan nodes. You probably have to identify actual output operator for a given plan node. `com.facebook.presto.sql.planner.PlanPrinter#getPlanNodeStats` contains code for gathering detailed input and output stats for a plan node